geo/geogfn: implement distance and dwithin for spheres using S2

otan opened the pull request:
This commit adds support for Distance and DWithin using spheres. This is
a wrapper around the S2 library, using their fancy ShapeIndex and
MinDistance edge finders to do the dirty work.

(I wish they supported giving me the two closest points so we can
completely re-use it for spheroids, but they do not.)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47983)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-:-M5hbrjZ8J3OxmncHflA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geogfn/distance.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-M5hbOlp0RGUtq98e9U0:-M5hbOlp0RGUtq98e9U1:b-6e498z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4bb22faced093423947af38b4670d52070d636/pkg/geo/geogfn/distance.go#L19)):*
> ```Go
> )
> 
> // wgs84SphereRadius is the radius if WGS84 was a sphere.
> ```

... is the radius in meters ... 
___
*[pkg/geo/geogfn/dwithin.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-M5hWfDuDTT03HPcb4ZT:-M5hWfDuDTT03HPcb4ZU:b-ewhfuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4bb22faced093423947af38b4670d52070d636/pkg/geo/geogfn/dwithin.go#L20)):*
> ```Go
> 
> // DWithin returns whether a is within distance d of b, i.e. Distance(a, b) <= d.
> func DWithin(a *geo.Geography, b *geo.Geography, distance float64) (bool, error) {
> ```

This and `Distance` are very similar, including the early termination logic (`Distance` terminates at 0). Can we have a single helper function for both that is something like:
```
func distanceWithLowerBound(a *geo.Geography, b *geo.Geography, distanceLowerBound float64) (reachedLowerBound bool, latestDistance float64, err error)
```
which we would call with 0 for `Distance` and the `DWithin` distance parameter.

Hmm, I guess not, since `IsDistanceLess()` is supposed to be much faster.
___
*[pkg/geo/geogfn/dwithin_test.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-M5hbiOvCq0X4JFvQQZu:-M5hbiOvCq0X4JFvQQZv:b-wtpf8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4bb22faced093423947af38b4670d52070d636/pkg/geo/geogfn/dwithin_test.go#L54)):*
> ```Go
> 				} else {
> 					for _, val := range []float64{
> 						tc.expectedSphereDistance + 0.01, // allow 1cm margin of error
> ```

can you add +0.02 here and -0.01 in the case below?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47983)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/distance.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-M5hbOlp0RGUtq98e9U0:-M5iUz721YS0gpv_NQbb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4bb22faced093423947af38b4670d52070d636/pkg/geo/geogfn/distance.go#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

... is the radius in meters ... 
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/dwithin.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-M5hWfDuDTT03HPcb4ZT:-M5iUzlNCT8RJxXTupKN:bgnxk1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4bb22faced093423947af38b4670d52070d636/pkg/geo/geogfn/dwithin.go#L20)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This and `Distance` are very similar, including the early termination logic (`Distance` terminates at 0). Can we have a single helper function for both that is something like:
```
func distanceWithLowerBound(a *geo.Geography, b *geo.Geography, distanceLowerBound float64) (reachedLowerBound bool, latestDistance float64, err error)
```
which we would call with 0 for `Distance` and the `DWithin` distance parameter.

Hmm, I guess not, since `IsDistanceLess()` is supposed to be much faster.
</blockquote></details>

Yeah nah couldn't really make anything more similar.
___
*[pkg/geo/geogfn/dwithin_test.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47983#-M5hbiOvCq0X4JFvQQZu:-M5iV1RX1U6eQYVbBbkV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b4bb22faced093423947af38b4670d52070d636/pkg/geo/geogfn/dwithin_test.go#L54)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add +0.02 here and -0.01 in the case below?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
geo/geogfn: implement distance and dwithin for spheres using S2

This commit adds support for Distance and DWithin using spheres. This is
a wrapper around the S2 library, using their fancy ShapeIndex and
MinDistance edge finders to do the dirty work.

(I wish they supported giving me the two closest points so we can
completely re-use it for spheroids, but they do not.)

Release note: None
